### PR TITLE
Prevent running Presto on Java versions prior to 1.8u60

### PR DESCRIPTION
**Pull request by yinghai**
This patch is to address issue: https://github.com/prestodb/presto/issues/5150

Test major version first, if major version equals to "1.8.0",
compare minor version with "60".

### CODE REVIEW

**Comment by cawallin** [comment]
Can you update the README along with this change? It will probably have infrastructure implications for folks running Presto, since the previous minimum version of Java was 1.8u40.

**Comment by yinghai** [comment]
OK.

**Comment by cawallin** [comment]
Thanks! Also, it would probably make sense to update presto-server-rpm at the same time. In this line, you need to change 40 to 60: https://github.com/prestodb/presto/blob/master/presto-server-rpm/src/main/rpm/preinstall#L34

**Comment by ArturGajowy** [comment]
I think some Travis config updates will be needed for the PR build to pass, as Travis uses an older JDK version:

```
$ jdk_switcher use oraclejdk8
Switching to Oracle JDK8 (java-8-oracle-amd64), JAVA_HOME will be set to /usr/lib/jvm/java-8-oracle-amd64
update-alternatives: error: no alternatives for apt
update-java-alternatives: jdk alternative does not exist: /usr/lib/jvm/java-8-oracle-amd64/bin/apt
$ java -Xmx32m -version
java version "1.8.0_51"
```

**Comment by cawallin** [comment]
Looking at the failures, the docker containers used to run the product tests will also need to be updated. Someone at Teradata will fix that and upload the new docker images -- we'll let you know when that's done.

**Comment by yinghai** [comment]
OK. Thanks. Is there an issue#?

**Comment by cawallin** [comment]
Yes, see the Travis USE_PRODUCT_TESTS failure:
```
Creating singlenode_presto-master_1
Attaching to singlenode_presto-master_1
presto-master_1 | Listening for transport dt_socket at address: 5005
presto-master_1 | Presto requires Java 1.8.0_60+ (found 1.8.0_40)
```

**Comment by yinghai** [comment]
Folks, what changes need to be done?

**Comment by cawallin** [comment]
I'm about to upload the new docker images now, and then the docker-compose.yml files in presto-product-tests/docker/[multinode|singlenode] need to be updated.

**Comment by cawallin** [comment]
docker-compose.yml is updated here: https://github.com/prestodb/presto/pull/5229

**Comment by yinghai** [comment]
Thanks, I'll update the yml soon.

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java [hunk 1]
```diff
@@ -36,9 +36,17 @@ private PrestoSystemRequirements() {}
 
     public static void verifyJvmRequirements()
     {
-        String specVersion = StandardSystemProperty.JAVA_SPECIFICATION_VERSION.value();
-        if ((specVersion == null) || (specVersion.compareTo("1.8") < 0)) {
-            failRequirement("Presto requires Java 1.8+ (found %s)", specVersion);
+        String javaVersion = StandardSystemProperty.JAVA_VERSION.value();
+        int sep = (javaVersion == null) ? -1 : javaVersion.lastIndexOf('_');
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: We don't abbreviate variable names, so name this `separator` or `index`.

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: This code can be simplified and the error message improved by failing early:
```java
String javaVersion = StandardSystemProperty.JAVA_VERSION.value();
if (javaVersion == null) {
    failRequirement("Java version not defined");
}
String[] versionParts = javaVersion.split("_");
if (versionParts.length != 2) {
    failRequirement("Java version has an unknown format: %s", javaVersion);
}
String majorVersion = versionParts[0];
String minorVersion = versionParts[1];

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java [hunk 2]
```diff
@@ -36,9 +36,17 @@ private PrestoSystemRequirements() {}
...
+        String majorVersion = null;
+        String minorVersion = null;
+        if (sep >= 0) {
+            majorVersion = javaVersion.substring(0, sep);
+            minorVersion = javaVersion.substring(sep + 1);
+        }
+        if ((majorVersion == null) || (minorVersion == null) || (majorVersion.compareTo("1.8.0") < 0) ||
+                (majorVersion.compareTo("1.8.0") == 0 && minorVersion.compareTo("60") < 0)) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: Comparing `60` as a string will break when an update higher than `99` is released. For example, the most recent (non-public) release of Java 7 is 7u101: http://www.oracle.com/technetwork/java/javase/documentation/javase7supportreleasenotes-1601161.html#R170_101

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java (hunk 1 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: Let's move the version check below the check for `JAVA_VENDOR`. Now that this is checking for a very specific version string format, it is likely to fail on a non-Oracle JVM with a confusion error message.

**Comment by yinghai** [comment]
Thanks, I'll update accordingly.

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java [hunk 3]
```diff
@@ -36,16 +36,33 @@ private PrestoSystemRequirements() {}
...
+        if (versionParts.length != 2) {
+            failRequirement("Java version has an unknown format: %s", javaVersion);
+        }
+        String majorVersion = versionParts[0];
+        int minorVersion;
+        try {
+            minorVersion = Integer.parseInt(versionParts[1]);
+        }
+        catch (Exception e) {
+            minorVersion = 0;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: Fail here directly:
```
failRequirement("Java minor version is not a number: %s", javaVersion);

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java [hunk 4]
```diff
@@ -36,16 +36,33 @@ private PrestoSystemRequirements() {}
...
+        }
+
+        if ((majorVersion == null) || (majorVersion.compareTo("1.8.0") < 0) ||
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: `majorVersion` can't be null

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java [hunk 5]
```diff
@@ -36,16 +36,33 @@ private PrestoSystemRequirements() {}
...
+                (majorVersion.compareTo("1.8.0") == 0 && minorVersion < 60)) {
+            failRequirement("Presto requires Java 1.8.0_60+ (found %s_%d)", majorVersion, minorVersion);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: Probably better to give the "marketing name" and print the exact version string (for debugging when someone reports an error):
```java
failRequirement("Presto requires Java 8u60+ (found %s)", javaVersion);

File: presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java [hunk 6]
```diff
@@ -36,16 +36,33 @@ private PrestoSystemRequirements() {}
...
+        String[] versionParts = javaVersion.split("_");
+        if (versionParts.length != 2) {
+            failRequirement("Java version has an unknown format: %s", javaVersion);
+        }
+        String majorVersion = versionParts[0];
+        int minorVersion;
+        try {
+            minorVersion = Integer.parseInt(versionParts[1]);
+        }
+        catch (Exception e) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/PrestoSystemRequirements.java]: This should catch `NumberFormatException`. Always be conservative when catching exceptions

**Comment by electrum** [comment]
Merged, thanks!

**Comment by cawallin** [comment]
This build still fails because the Travis Java 8 version is 1.8.0_52. Also, since the Kerberos tests were merged at basically the same time as this patch, we still need to update the docker images for those.

**Comment by yinghai** [comment]
I added a new patch to address more review comments above. How to update Travis?

**Comment by cawallin** [comment]
There is a workaround in this Travis issue that you can try: https://github.com/travis-ci/travis-ci/issues/3259.

Alternatively, you (or @electrum or someone else) can reach out to the Travis team to push them on updating the Java version.

**Pull request closed without merging**